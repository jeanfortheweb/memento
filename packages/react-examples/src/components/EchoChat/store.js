import { Record, List } from 'immutable';
import { Store } from '@memento/store';
import createMade, { set, push } from '@memento/made';
import createSnitch, { listen } from '@memento/snitch';
import createSupervisor, { sequence } from '@memento/supervisor';
import createHotline, {
  connect,
  disconnect,
  close,
  open,
  send,
  receive,
} from '@memento/hotline';

class Message extends Record({
  date: 0,
  username: '',
  text: '',
}) {}

class State extends Record({
  username: '',
  text: '',
  isConnected: false,
  messages: List(),
}) {}

export const setUsername = event => set('username', event.target.value);
export const getUsername = state => state.username;

export const setText = event => set('text', event.target.value);
export const getText = state => state.text;

export const sendChatMessage = (username, text) => () => {
  const sendMessage = send({
    username,
    text,
    date: Date.now(),
  });

  const clearText = set('text', '');

  return sequence(sendMessage, clearText);
};

export const getMessages = state => state.messages;
export const isConnected = state => state.isConnected;
export const connectToChat = () => connect('ws://echo.websocket.org');
export const disconnectFromChat = () => disconnect();

const registerListeners = () => {
  const onConnect = listen(open, () => set('isConnected', true));
  const onReceive = listen(receive, ({ data }) => push('messages', new Message(data)));
  const onClose = listen(close, () =>
    sequence(set('isConnected', false), set('messages', List())),
  );

  return sequence(onConnect, onClose, onReceive);
};

const store = new Store(new State(), [
  createMade(),
  createHotline(),
  createSnitch(),
  createSupervisor(),
]);

store.assign(registerListeners());

export default store;
