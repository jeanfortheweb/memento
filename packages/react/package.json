{
  "name": "@memento/react",
  "version": "0.0.1",
  "license": "MIT",
  "main": "es5/index.js",
  "types": "es5/index.d.ts",
  "scripts": {
    "test": "jest",
    "coverage": "jest --coverage",
    "prepublish": "tsc --p ./tsconfig.publish.json"
  },
  "publishConfig": {
    "access": "public"
  },
  "keywords": ["flux", "state management"],
  "author": "jeanfortheweb",
  "devDependencies": {
    "@types/jest": "^22.1.1",
    "@types/react-test-renderer": "^16.0.0",
    "@types/react": "^16.0.36",
    "jest": "^22.1.4",
    "ts-jest": "^22.0.2",
    "typescript": "^2.7.1"
  },
  "dependencies": {
    "@memento/store": "^0.0.1",
    "immutable": "4.0.0-rc.9",
    "react": "^16.2.0",
    "react-test-renderer": "^16.2.0"
  },
  "jest": {
    "bail": true,
    "transform": {
      "^.+\\.tsx?$": "ts-jest"
    },
    "moduleFileExtensions": ["ts", "tsx", "js"],
    "testRegex": "\\.spec\\.tsx?$",
    "mapCoverage": true
  }
}
