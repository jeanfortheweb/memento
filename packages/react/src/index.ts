import View, { Props as ViewProps } from './components/View';
import Trigger, { Props as TriggerProps } from './components/Trigger';

export { View, ViewProps, Trigger, TriggerProps };
