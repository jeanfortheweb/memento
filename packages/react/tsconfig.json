{
  "compilerOptions": {
    "module": "commonjs",
    "target": "es5",
    "lib": ["es6", "dom"],
    "outDir": "./es5",
    "declaration": true,
    "strict": true,
    "noImplicitAny": false,
    "noImplicitThis": false,
    "strictNullChecks": true,
    "jsx": "react",
    "forceConsistentCasingInFileNames": true,
    "noUnusedLocals": true,
    "skipLibCheck": true
  },
  "include": ["src", "node_modules/@types"]
}
