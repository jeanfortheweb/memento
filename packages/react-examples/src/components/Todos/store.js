import { Observable, withLatestFrom, flatMap } from '@reactivex/rxjs';
import { Record, List } from 'immutable';
import shortid from 'shortid';
import { Store } from '@memento/store';
import createSupervisor, { sequence, when } from '@memento/supervisor';
import createMade, { push, merge, update, set } from '@memento/made';
import createSnitch, { listen, unlisten } from '@memento/snitch';
import createClerk, { SaveMode, LoadMode, Target } from '@memento/clerk';

// state.
export class Todo extends Record({
  id: null,
  date: Date.now(),
  text: 'Enter a todo text',
  done: false,
}) {}

export class State extends Record({
  todos: List(),
  text: '',
  filter: 'ALL',
}) {}

// task creators.
export const toggleTodo = todo => () =>
  update('todos', todo, todo.set('done', !todo.done));

export const setTodoText = event => set('text', event.target.value);
export const setFilter = value => () => set('filter', value.toUpperCase());

export const addTodo = text => () => {
  const pushTodo = push('todos', new Todo({ id: shortid.generate(), text }));
  const clearText = set('text', '');

  return sequence(pushTodo, clearText);
};

// selectors.
export const getTodos = state => state.todos;
export const getTodoText = state => state.text;
export const getFilter = state => state.filter;

// create the store with required workers.
const store = new Store(new State(), [
  createSnitch(),
  createMade(),
  createSupervisor(),
  createClerk({
    name: 'todos',
    target: Target.Local,
    save: SaveMode.Auto,
    load: LoadMode.Auto,
    path: 'todos',
    reviver: (key, sequence) => {
      if (typeof key === 'number') {
        return new Todo(sequence);
      }

      return sequence.toList();
    },
  }),
]);

store.assign(
  when(
    state => state.todos.size === 0,
    () => sequence(addTodo('Add more features')(), addTodo('Update documentation')()),
  ),
);

export default store;
