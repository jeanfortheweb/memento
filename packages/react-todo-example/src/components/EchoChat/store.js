import { Record, List } from 'immutable';
import { Store } from '@memento/store';
import createMade, { set, push } from '@memento/made';
import createSnitch, { listen } from '@memento/snitch';
import createSequencer, { sequence } from '@memento/sequencer';
import createHotline, {
  connect,
  disconnect,
  close,
  open,
  send,
  received,
} from '@memento/hotline';

class Message extends Record({
  date: 0,
  username: '',
  text: '',
}) {}

class State extends Record({
  username: '',
  text: '',
  isConnected: false,
  messages: List(),
}) {}

export const setUsername = event => set('username', event.target.value);
export const getUsername = state => state.username;

export const setText = event => set('text', event.target.value);
export const getText = state => state.text;

export const sendChatMessage = (username, text) => () => {
  const sendMessage = send(
    new Message({
      username,
      text,
      date: Date.now(),
    }).toJS(),
  );

  const clearText = set('text', '');

  return sequence(sendMessage, clearText);
};

export const getMessages = state => state.messages;
export const isConnected = state => state.isConnected;
export const connectToChat = () => connect('ws://echo.websocket.org');
export const disconnectFromChat = () => disconnect();

const registerListeners = () => {
  const onConnect = listen(open, () => set('isConnected', true));
  const onReceive = listen(received, ({ data }) => push('messages', new Message(data)));
  const onClose = listen(close, () =>
    sequence(set('isConnected', false), set('messages', List())),
  );

  return sequence(onConnect, onClose, onReceive);
};

const store = new Store(new State(), [
  createMade(),
  createHotline(),
  createSnitch(),
  createSequencer(),
]);

store.assign(registerListeners());

export default store;
