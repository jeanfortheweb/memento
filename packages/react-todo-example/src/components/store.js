import { Record } from 'immutable';
import { Store } from '@memento/store';
import createMade, { set } from '@memento/made';

export class State extends Record({
  active: 'Todos',
}) {}

export const getActive = state => state.active;
export const setActive = name => () => set('active', name);

export default new Store(new State(), [createMade()]);
