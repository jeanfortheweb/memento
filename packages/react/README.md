# Memento/Made - State Manipulation Tools

[![Build Status](https://travis-ci.org/jeanfortheweb/memento.svg?branch=master)](https://travis-ci.org/jeanfortheweb/memento) [![Maintainability](https://api.codeclimate.com/v1/badges/5494041ca69fd977cae6/maintainability)](https://codeclimate.com/github/jeanfortheweb/memento/maintainability) [![Test Coverage](https://api.codeclimate.com/v1/badges/5494041ca69fd977cae6/test_coverage)](https://codeclimate.com/github/jeanfortheweb/memento/test_coverage)

This is the official React binding for Memento. It provides declarative components to bind your presentational components to your memento state and task creators in a flux like pattern. Unlike most libraries, these components are not created by decorators and do not generate hidden code. Instead, they make use of the render props pattern to make the code fully declarative, conflict free and easy to follow.

## Documentation

You can find the memento documentation at [memento.js.org](http://memento.js.org).

## Examples

The [react examples](https://github.com/jeanfortheweb/memento/tree/master/packages/react-examples) are our playground for new features and showing off alot of mementos power.

Here is a simple example showing a message which gets recomputed to uppercase and then displayed.

```js
import { Store } from '@memento/store';
import { View } from '@memento/react';
import { Record } from 'immutable';
import React from 'react';
import ReactDOM from 'react-dom';

class State extends Record({
  message: 'Hello World',
}) {}

const store = new Store(new State(), []);

const App = () => (
  <View
    store={store}
    message={state => state.message}
    compute={props => props.message.toUpperCase()}
  >
    {yelledMessage => <div>{yelledMessage}</div>}
  </View>
);

ReactDOM.render(<App />, document.getElementById('#app'));
```

## Installation

To use Memento and React, you'll need to install at least:

```sh
yarn install @memento/store @memento/react @memento/made immutable react react-dom
```
