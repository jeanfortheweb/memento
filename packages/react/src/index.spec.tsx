import { Store, Selector } from '@memento/store';
import * as renderer from 'react-test-renderer';
import { Record } from 'immutable';
import { Projector } from './';
import * as React from 'react';

describe('@memento/react', () => {
  interface StateShape {
    username: string;
    age: number;
  }

  interface StateShape2 {
    username2: string;
    age2: number;
  }

  class State extends Record<StateShape>({
    username: 'Foo',
    age: 1337,
  }) {}

  class State2 extends Record<StateShape2>({
    username2: 'Foo',
    age2: 1337,
  }) {}

  const store = new Store(new State(), []);
  const StringProjector = Projector.for<State, string>();

  const getUsername: Selector<State, string> = state => state.username;

  test('initial render as expected', () => {
    const instance = <StringProjector store={store} selector={getUsername} />;

    const component = renderer.create(instance);

    const tree = component.toJSON();

    expect(tree).toMatchSnapshot();
  });
});
