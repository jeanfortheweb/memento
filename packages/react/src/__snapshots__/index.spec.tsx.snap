// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`@memento/react initial render as expected 1`] = `"Foo"`;
